piptools/scripts/compile.py

```diff
@@ -291,6 +305,27 @@ def _determine_linesep(
     help="Specify a package to consider unsafe; may be used more than once. "
     f"Replaces default unsafe packages: {', '.join(sorted(UNSAFE_PACKAGES))}",
 )
+@click.option(
+    "--build-distribution",
+    "build_distributions",
+    multiple=True,
+    type=click.Choice(("editable", "sdist", "wheel")),
+    help="Name of a distribution to install build dependencies for; may be used more than once. "
+    "Static dependencies declared in pyproject.toml will be included as well.",
+)
+@click.option(
+    "--all-build-distributions",
```

Comment:
Staring at this, I realized that his name does not mention build dependencies. It needs to be something like `--all-build-deps` or smth.

Reply:
My thinking was that e.g. "sdist" is not a build dependency (a build dependency would be something like cffi) so calling it something like `--build-dep` is not a perfect fit. On the other hand the other options make more sense when following this pattern, they could be called something like `--all-build-deps` and `--only-build-deps`.
 
Since it wasn't a perfect fit I looked at what terminology is used elsewhere and noted that the argument to `get_requires_for_build` is called "distribution".

Another option I can think of is `--build-deps-for` (the other options could still be called `--all-build-deps` and `--only-build-deps`).

Either way I don't have a strong preference so what do you prefer?

Reply:
I'm not sure we're talking about the same things. Distribution packages are archives published to PyPI (like wheels and tarballs/sdists). And this option is about controlling build dependencies which are declared by `[build-system].requires` and the `get_requires_for_build_*()` hooks.

Reply:
> Distribution packages are archives published to PyPI (like wheels and tarballs/sdists)

Agreed.

> And this option is about controlling build dependencies which are declared by [build-system].requires and the get_requires_for_build_*() hooks.

Agreed. The complete list of places where build dependencies can currently be declared is
* `[build-system].requires`
* `get_requires_for_build_editable`
* `get_requires_for_build_sdist`
* `get_requires_for_build_wheel`

Do you have a suggestion for what these place-where-build-dependencies-are-declared should be called?
The `build` package calls the latter three `distribution`:

```python
    def get_requires_for_build(self, distribution: str, config_settings: ConfigSettingsType | None = None) -> set[str]:
        """
        Return the dependencies defined by the backend in addition to
        :attr:`build_system_requires` for a given distribution.

        :param distribution: Distribution to get the dependencies of
            (``sdist`` or ``wheel``)
        :param config_settings: Config settings for the build backend
        """
```

Either way, I think we can avoid using the name on the CLI if we go with the following three option names
* `--build-deps-for` (Akin to `--extras`. Acceptable values to this option are `editable`, `sdist`, and `wheel`.)
* `--all-build-deps` (Akin to `--all-extras`)
* `--build-deps-only`

What do you think?

Reply:
> Do you have a suggestion for what these place-where-build-dependencies-are-declared should be called?
> The `build` package calls the latter three `distribution`:
> 
> ```python
>     def get_requires_for_build(self, distribution: str, config_settings: ConfigSettingsType | None = None) -> set[str]:
>         """
>         Return the dependencies defined by the backend in addition to
>         :attr:`build_system_requires` for a given distribution.
> 
>         :param distribution: Distribution to get the dependencies of
>             (``sdist`` or ``wheel``)
>         :param config_settings: Config settings for the build backend
>         """
> ```

The build targets are distributions indeed. Though, our naming issue is with the build deps for said targets.

> Either way, I think we can avoid using the name on the CLI if we go with the following three option names
> * `--build-deps-for` (Akin to `--extras`. Acceptable values to this option are `editable`, `sdist`, and `wheel`.)

Any idea how would one ask for multiple subtypes? Multiple arg use or one arg and a comma-separated value? I wonder what https://clig.dev has to say on the topic..

> * `--all-build-deps` (Akin to `--all-extras`)
> * `--build-deps-only`
> 
> What do you think?

That sounds reasonable.

Reply:
Okay, marking this as resolved.